Admin endpoint to modify user tier

yenicelik opened the pull request:
#16 

- Add admin endpoints + table
- Add endpoint to modify a users tier as an admin
- Add script to create an admin.

An admin can use the CI tool to update a user's role.

WyseNynja commented on web3_proxy/src/frontend/users.rs:
```diff
@@ -853,3 +855,135 @@ pub async fn user_stats_detailed_get(
+
+    // Modify the user-role ...
+    // Check if this use has admin privileges ...
+    let user_role: secondary_user::Model = secondary_user::Entity::find()
```
`secondary_user` is going to be used for something else. Users will be able to add other users who can view their stats and administrate their account.

Site-wide admins will be on a table that still needs to be created.

This should be a query on the `user` table to find the user that the admin is trying to modify.

yenicelik replied:
added an admin table (rn without "description", i can add it now, or also at a later migration). as you prefer.

WyseNynja commented on web3_proxy/src/frontend/users.rs:
```diff
@@ -853,3 +855,135 @@ pub async fn user_stats_detailed_get(
+
+    // TODO: Make a single query, where you retrieve the user, and directly from it the secondary user (otherwise we do two jumpy, which is unnecessary)
+    // get the user id first. if it is 0, we should use a cache on the app
+    let user_id = get_user_id_from_params(&mut redis_conn, &db_conn, &db_replica, bearer, &params).await?;
```
let's call this `admin_id` and then add a query here that checks that the `admin_id` is in an `admin` table

yenicelik replied:
I called it caller_id (which is later resolved to an admin, if he is truthfully an admin). can change it to admin_id if you prefer

WyseNynja replied:
caller_id is good

WyseNynja commented on web3_proxy/src/frontend/users.rs:
```diff
@@ -853,3 +855,135 @@ pub async fn user_stats_detailed_get(
+
+    // Get both the user and user role
+    let user: user::Model = user::Entity::find_by_id(user_id)
+        .one(&db_conn)
```
this should use `db_replica`

yenicelik replied:
not entirely sure, but what is the (practical) difference to why we have both a db_replica, and a db_conn? speed?

WyseNynja replied:
db_replica is read only and replicated all over the world. It's much faster to query since everything doesn't have to go all the way to Ohio. 

If you are just reading the data, use replica.

If you are reading the data to modify and then write it or are writing brand new data, use the main db.

WyseNynja commented on web3_proxy/src/frontend/mod.rs:
```diff
@@ -85,6 +85,7 @@ pub async fn serve(port: u16, proxy_app: Arc<Web3ProxyApp>) -> anyhow::Result<()
             get(users::user_stats_aggregated_get),
         )
         .route("/user/stats/detailed", get(users::user_stats_detailed_get))
+        .route("/user/modify_role", get(users::admin_change_user_roles))
```
let's put this at `/admin/modify_role`. Regular users should not have access to do this.

yenicelik commented:
- Added removing user's login cache (entry in Login). 
- Added script to upgrade a user to an admin through CLI
- Create Admin table.

@WyseNynja again, would be great to have some guidance reg. tests (where and how these should be added).

WyseNynja commented:
reminder: add a test for this new endpoint

yenicelik commented:
Will close this PR as the changes are included in https://github.com/llamanodes/web3-proxy/pull/20
